### DISCUSSION TITLE
Multi function gathered result / Aiohttp - use Client with event loop inside view to get result

### DISCUSSION THREAD

**Discussion by js-compilatrum** [fastapi/fastapi#7450] [Questions] [answered]
### First check

* [x ] I used the GitHub search to find a similar issue and didn't find it.
* [ x] I searched the FastAPI documentation, with the integrated search.
* [ x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

I want use code:

```
async def fetch(session: ClientSession, url: str) -> str:
    async with session.get(url) as resp:
        return await resp.text()


async def main(urls: List[str]):
    async with aiohttp.ClientSession() as session:
        tasks = [asyncio.ensure_future(fetch(session, url)) for url in urls]
        response_data = await asyncio.gather(*tasks)
        return response_data


def get_code(urls: list):
    data = asyncio.run(main(urls))
    return data

```

to create API call like this:

```
router = APIRouter()
@router.get("/sourcecode/from-url")
async def online_sources(link: AnyUrl):
    data = get_code(list(link))
      return {'data': data}
```

but this code create internal server error:
```
  File "/home/pi/.local/lib/python3.7/site-packages/nest_asyncio.py", line 11, in apply
    raise ValueError('Can\'t patch loop of type %s' % type(loop))
ValueError: Can't patch loop of type <class 'uvloop.Loop'>
```
### Additional context

Targer platform is Raspberry PI 4 with Python 3.7.3 (Rasbian OS)

### Perspective
My target is create NLP service where using API call is possible get response from bunch of urls and process them using NLP. My code outside Fast API to get source code and result of externatl calls working, but inside Fast API I have not idea how coded it to makes them working.

For single call I found intergration:
[Single call aiohtpp](https://github.com/raphaelauv/fastAPI-aiohttp-example/blob/master/src/fastAPI_aiohttp/fastAPI.py)

but I want inside Fast API parts when one function gathers multitime result (async gather result) and try find patter to achieve something similar to code provided above.  Provided code is with single URL, but I want modified design to use multi URL passed by this API point inside aiohttp Client. I have not idea how move forward and resolve this.

Best regards

**Comment by dbanty** [comment]
Not at a computer to test right now but I think the issue is that FastAPI is already running asyncio for you. So instead of trying to run your coroutine with `asyncio.run` in `get_data` you should just be awaiting your coroutine. 

So try this:

```python
router = APIRouter()
@router.get("/sourcecode/from-url")
async def online_sources(link: AnyUrl):
    data = await main(list(link)) # this is the change
    return {'data': data}
```

**Comment by js-compilatrum** [comment]
I'm looking pattern to use async code with Fast API. Your suggestion is good when only **one** website is process (one functions with params), because still I have asyncio.gather which is not compatible as you suggest me. My target is:

1. API call
2. fill from function data from call
3. run the same function multitimes with other data
4. get / put somewhere result of operation to use with other API calls.

**Comment by raphaelauv** [comment, marked as answer]
I wrote this -> [Single call aiohtpp](https://github.com/raphaelauv/fastAPI-aiohttp-example/blob/master/src/fastAPI_aiohttp/fastAPI.py)

This is an example for multiple calls

```python
@app.get('/endpoint')
async def endpoint():
    url = "http://localhost:8080/test"

    with aioresponses() as mock_server:  # mock answer , remove in real
        mock_server.post(url=url, status=200, body=json.dumps({"succes": 1}))
        mock_server.post(url=url, status=200, body=json.dumps({"succes": 2}))

        async_calls: List[Coroutine] = list()  # store all async operations

        async_calls.append(SingletonAiohttp.query_url(url))
        async_calls.append(SingletonAiohttp.query_url(url))

        all_results: List[Tuple] = await asyncio.gather(*async_calls)  # wait for all async operations

    return 1
```

code with test is [there](https://github.com/raphaelauv/fastAPI-aiohttp-example/blob/b5b55592807ed2f8218511a986c6519babc66d11/src/fastAPI_aiohttp/fastAPI.py#L77)

**Comment by raphaelauv** [comment]
@js-compilatrum
can you close the issue

**Comment by js-compilatrum** [comment]
Ops! My mistake. Thank you for your help.

**Comment by tiangolo** [comment]
Thanks for the help here @dbanty and @raphaelauv ! 👏  🙇 

Thanks for reporting back and closing the issue @js-compilatrum  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
